src/fsharp/Microsoft.DotNet.DependencyManager/NativeDllResolveHandler.fs

```diff
@@ -98,28 +96,47 @@ type NativeDllResolveHandlerCoreClr (nativeProbingRoots: NativeResolutionProbe)
     let eventInfo = typeof<AssemblyLoadContext>.GetEvent("ResolvingUnmanagedDll")
     let handler = Func<Assembly, string, IntPtr> (_resolveUnmanagedDll)
 
-    do if not (isNull eventInfo) then eventInfo.AddEventHandler(AssemblyLoadContext.Default, handler)
+    do
+        if not (isNull eventInfo) then
+            eventInfo.AddEventHandler(AssemblyLoadContext.Default, handler)
 
     interface IDisposable with
         member _x.Dispose() =
             if not (isNull eventInfo) then
                 eventInfo.RemoveEventHandler(AssemblyLoadContext.Default, handler)
             ()
 
-#endif
-
-type NativeDllResolveHandler (_nativeProbingRoots: NativeResolutionProbe) =
-
+type NativeDllResolveHandler (nativeProbingRoots: NativeResolutionProbe) =
+    let addedPaths = ConcurrentBag<string>()
     let handler:IDisposable option =
-#if NETSTANDARD
         if isRunningOnCoreClr then
-            Some (new NativeDllResolveHandlerCoreClr(_nativeProbingRoots) :> IDisposable)
+            Some (new NativeDllResolveHandlerCoreClr(nativeProbingRoots) :> IDisposable)
         else
-#endif
             None
 
+    member internal _.RefreshPathsInEnvironment() =
+        for probePath in (nativeProbingRoots.Invoke()) do
+            let path =
+                let p = Environment.GetEnvironmentVariable("PATH")
+                if not(p.EndsWith(";", StringComparison.OrdinalIgnoreCase)) then p + ";"
+                else p
+
+            if not (probePath.Contains(path + ";")) then
+                Environment.SetEnvironmentVariable("PATH", path + probePath + ";")
+                addedPaths.Add path
+
     interface IDisposable with
         member _.Dispose() =
             match handler with
             | None -> ()
             | Some handler -> handler.Dispose()
+
+            for probePath in addedPaths do
+                let path =
+                    let p = Environment.GetEnvironmentVariable("PATH")
+                    if not(p.EndsWith(";", StringComparison.OrdinalIgnoreCase)) then p + ";"
+                    else p
+
+                Environment.SetEnvironmentVariable("PATH", path.Replace(probePath + ";", ""))
+                let mutable _p = null
+                addedPaths.TryTake(ref _p) |> ignore
```

Comment:
I'm having a hard time tracking what these new bits do.  Can you refactor to something like `let addToPath(existingPath: string, extraDir: string) -> string`/`let removeFromPath(existingPath: string, removeDir: string) -> string` and add unit tests?  Then these locations simply pipe in to and out of `%PATH%`.

Reply:
addedPaths is the collection of paths that have been added.  Collected, so that they can be removed in dispose.  But I can refactor some stuff yes.